cpp/src/parquet/util/crypto.cc

```diff
@@ -210,24 +224,31 @@ int ctr_encrypt(const uint8_t* plaintext, int plaintext_len, uint8_t* key, int k
   }
 
   // Encryption
-  if (1 != EVP_EncryptUpdate(cipher.get(), ciphertext + ctrIvLen, &len, plaintext,
+  if (1 != EVP_EncryptUpdate(cipher.get(), ciphertext + bufferSizeLen + ctrIvLen, &len, plaintext,
                              plaintext_len)) {
     throw ParquetException("Failed encryption update");
   }
 
   ciphertext_len = len;
 
   // Finalization
-  if (1 != EVP_EncryptFinal_ex(cipher.get(), ciphertext + ctrIvLen + len, &len)) {
+  if (1 != EVP_EncryptFinal_ex(cipher.get(), ciphertext + bufferSizeLen + ctrIvLen + len, &len)) {
     throw ParquetException("Failed encryption finalization");
   }
 
   ciphertext_len += len;
 
-  // Copying the IV ciphertext
-  std::copy(iv, iv + ctrIvLen, ciphertext);
-
-  return ctrIvLen + ciphertext_len;
+  // Copying the buffer size and IV to ciphertext
+  int bufferSize = ctrIvLen + ciphertext_len;
+  uint8_t bufferSizeArray [bufferSizeLen];
+    bufferSizeArray[3] = 0xff & (bufferSize >> 24);
+    bufferSizeArray[2] = 0xff & (bufferSize >> 16);
+    bufferSizeArray[1] = 0xff & (bufferSize >> 8);
+    bufferSizeArray[0] = 0xff & (bufferSize);
+  std::copy(bufferSizeArray, bufferSizeArray + bufferSizeLen, ciphertext);
+  std::copy(iv, iv + ctrIvLen, ciphertext + bufferSizeLen);
+
+  return bufferSize;
```

Comment:
Now the bufferSize will have more 4 bytes for bufferSizeLen, then I think now we should `return` bufferSize + bufferSizeLen`? or the caller will have to add 4 bytes into bufferSize itself?

Reply:
@thamht4190 You are right, it should be + sizeLen (or rather, bufferSize should include this addition). I'll update the code.